crate_universe/src/utils/starlark/label.rs

```diff
@@ -11,60 +11,174 @@ use serde::{Deserialize, Serialize, Serializer};
+    }
+
+    #[cfg(test)]
+    pub fn repository(&self) -> Option<&str> {
+        match &self.package {
+            AbosluteOrRelativePackage::Relative => None,
+            AbosluteOrRelativePackage::Absolute { repository, .. } => repository.as_deref(),
+        }
+    }
+
+    pub fn package(&self) -> Option<&str> {
+        match &self.package {
+            AbosluteOrRelativePackage::Relative => None,
+            AbosluteOrRelativePackage::Absolute { package, .. } => package.as_deref(),
+        }
+    }
+
+    pub fn target(&self) -> &str {
+        self.target.as_str()
+    }
 }
 
 impl FromStr for Label {
     type Err = anyhow::Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let re = Regex::new(r"^(@@?[\w\d\-_\.]*)?/{0,2}([\w\d\-_\./+]+)?(:([\+\w\d\-_\./]+))?$")?;
+        let re = Regex::new(r"^(@@?[\w\d\-_\.]*)?(//)?([\w\d\-_\./+]+)?(:([\+\w\d\-_\./]+))?$")?;
         let cap = re
             .captures(s)
             .with_context(|| format!("Failed to parse label from string: {s}"))?;
 
-        let repository = cap
-            .get(1)
-            .map(|m| m.as_str().trim_start_matches('@').to_owned());
+        let repository = cap.get(1).map(|m| m.as_str().trim_start_matches('@'));
+
+        let is_absolute = cap.get(2).is_some();
+
+        let package = cap.get(3).map(|m| m.as_str());
+
+        let target = cap.get(5).map(|m| m.as_str());
+
+        match (repository, is_absolute, package, target) {
+            // Relative
+            (None, false, None, Some(target)) => Ok(Self {
+                package: AbosluteOrRelativePackage::Relative,
+                target: target.to_owned(),
+            }),
+
+            // Relative (Implicit Target which regex identifies as Package)
+            (None, false, Some(package), None) => Ok(Self {
+                package: AbosluteOrRelativePackage::Relative,
+                target: package.to_owned(),
+            }),
+
+            // Absolute (Full)
+            (Some(repository), true, Some(package), Some(target)) => Ok(Self {
+                package: AbosluteOrRelativePackage::Absolute {
+                    repository: Some(repository.to_owned()),
+                    package: Some(package.to_owned()),
+                },
+                target: target.to_owned(),
+            }),
+
+            // Absolute (Repository)
+            (Some(repository), _, None, None) => Ok(Self {
+                package: AbosluteOrRelativePackage::Absolute {
+                    repository: Some(repository.to_owned()),
+                    package: None,
+                },
+                target: repository.to_owned(),
+            }),
+
+            // Absolute (Package)
+            (None, true, Some(package), None) => Ok(Self {
+                package: AbosluteOrRelativePackage::Absolute {
+                    repository: None,
+                    package: Some(package.to_owned()),
+                },
+                target: package.to_owned(),
```

Comment:
This should just be the basename of the package:

`foo/bar` expands to `@//foo/bar:bar` not `@//foo/bar:foo/bar`

(And it should be an error if the target isn't specified and the package is empty)

Reply:
Adding a test for invalid input of `":"`.

Reply:
I believe the only case where you have an empty package but no target would be `//` (which is covered under `invalid_empty()`).  Something like `@repo//` would be an empty package and empty target but will default to `repo`.